zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+            region.assign_advice(
+                || "prev_hash",
+                self.raw_public_inputs,
+                offset,
+                || Ok(prev_hash),
+            )?;
+            raw_pi_vals[offset] = prev_hash;
+            offset += 1;
+        }
+
+        Ok(chain_id_cell)
+    }
+
+    /// Assigns the extra fields (not in block or tx tables):
+    ///   - state root
+    ///   - parent block hash
+    /// to the raw_public_inputs column and stores a copy in a
+    /// vector for computing RLC(raw_public_inputs).
+    #[allow(clippy::type_complexity)]
+    fn assign_extra_fields(
+        &self,
+        region: &mut Region<'_, F>,
+        extra: ExtraValues,
+        randomness: F,
+        raw_pi_vals: &mut Vec<F>,
+    ) -> Result<(AssignedCell<F, F>, AssignedCell<F, F>), Error> {
+        let mut offset = BLOCK_LEN;
+        // block hash
+        // let block_hash = rlc(extra.block_hash.to_fixed_bytes(), randomness);
+        // region.assign_advice(
+        //     || "block.hash",
+        //     self.raw_public_inputs,
+        //     offset,
+        //     || Ok(block_hash),
+        // )?;
+        // raw_pi_vals[offset] = block_hash;
+        // offset += 1;
+
+        // block root
+        let state_root = rlc(extra.state_root.to_fixed_bytes(), randomness);
+        let state_root_cell = region.assign_advice(
+            || "state.root",
+            self.raw_public_inputs,
+            offset,
+            || Ok(state_root),
+        )?;
+        raw_pi_vals[offset] = state_root;
+        offset += 1;
+
+        // parent block hash
+        let parent_block_hash = rlc(extra.parent_block_hash.to_fixed_bytes(), randomness);
+        let parent_block_hash_cell = region.assign_advice(
+            || "parent_block.hash",
+            self.raw_public_inputs,
+            offset,
+            || Ok(parent_block_hash),
+        )?;
+        raw_pi_vals[offset] = parent_block_hash;
+        Ok((state_root_cell, parent_block_hash_cell))
+    }
+
+    /// Assign `rpi_rlc_acc` and `rand_rpi` columns
+    #[allow(clippy::type_complexity)]
+    fn assign_rlc_pi(
+        &self,
+        region: &mut Region<'_, F>,
+        rand_rpi: F,
+        raw_pi_vals: Vec<F>,
+    ) -> Result<(AssignedCell<F, F>, AssignedCell<F, F>), Error> {
+        // TODO account for extra empty tx row if needed
+        let circuit_len = PiCircuitConfig::<F, MAX_TXS, MAX_CALLDATA>::circuit_len();
+        assert_eq!(circuit_len, raw_pi_vals.len());
+
+        // Last row
+        let offset = circuit_len - 1;
+        let mut rpi_rlc_acc = raw_pi_vals[offset];
+        region.assign_advice(
+            || "rpi_rlc_acc",
+            self.rpi_rlc_acc,
+            offset,
+            || Ok(rpi_rlc_acc),
+        )?;
+        region.assign_advice(|| "rand_rpi", self.rand_rpi, offset, || Ok(rand_rpi))?;
+        self.q_end.enable(region, offset)?;
+
+        // Next rows
+        for offset in (1..circuit_len).rev().skip(1) {
```

Comment:
```suggestion
        for offset in (1..circuit_len-1).rev() {
```
Maybe this is more clear?

Reply:
Yep